### PR TITLE
Feature/rivan 29482 update docker workflow steps

### PR DESCRIPTION
**Pull request by RollerKnobster**
## Description

This PR splits the existing `main.yml` Github workflow into CI and Docker related ones, also updating them with the usage of actions instead of direct bash commands.

[Azure DevOps Link](https://dev.azure.com/EquinorASA/OMNIA%20Prevent/_workitems/edit/29482)

## Type of change

Mark the relevant options (`[x]` will mark checkbox as checked).
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [x] CI related changes
- [x] Documentation content changes
- [ ] Tests
- [ ] Other

# Checklist:

- [x] I have made corresponding changes to the documentation / README (if needed)
- [ ] I have added tests that prove my fix is effective or that my feature works (if needed)

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,174 @@
+name: Docker
+
+on:
+  pull_request:
+    branches:
+      - master
```
Comment (devProdigy) [inline .github/workflows/docker.yml:6]: will it build docker image for each PR?
if yes, then as I told on the call - cause of the long time of build I'm against building docker on PRs, only for master branch (after merge). It'll slow down development.
This is not the ideal solution and should be changed after moving Gordo cliet to separate library.

File: README.md [hunk 2]
```diff
@@ -221,16 +222,51 @@ Do not forget update already installed packages:
...
+full version for a given commit during a pull request (`v1.0.1dev5`).
+
+The format of each part is configurable and may be changed to one's likings. For our purposes, they're set to `<major>.<minor>.<patch>`
+for the main part, `dev` for the delimiter part, and `<increment>` for the increment part. The "increment" stands for the number of commits for a specific version
+since the last commit that could set the highest version, starting from either the latest release tag or from the start of the log tree.
+
+A link to the GitHub release creation page is generated in the outputs of the Semantic Version step of Docker CI workflow,
+setting the tag to follow the format `<tag_prefix><version>`. `<tag_prefix>` is also a configurable variable, set to `v` for our purposes.
+
+Many examples and details on the topics can be found by following the links, but in general, every commit should follow a couple of rules:
```
Comment (devProdigy) [inline README.md:241]: sorry, I was confused what links?)
```suggestion
Many examples and details on the topics can be found by following the links above, but in general, every commit should follow a couple of rules:
```

File: README.md [hunk 3]
```diff
@@ -221,16 +222,51 @@ Do not forget update already installed packages:
...
+
+BREAKING CHANGE: `extends` key in config file is now used for extending other config files
+```
+
+Given this is the first commit since the release ref, the version after merging shall be set as `2.0.0dev0`.
+This makes sense due to the fact that a breaking change is present in a commit.
+A link to make a release with tag `v2.0.0` is going to be created.
+
+After pushing to master or creating a tag, the new image with the appropriate version is going to be 
+created and pushed to the registry.
```
Comment (devProdigy) [inline README.md:260]: Sorry, didn't understand this part.
So what is the common logic of the CI-docker-push?
Maybe some short summary/example could be added to the beginning "### Versioning" section?
Like:
```
1. create PR with `BREAKING FEATURE:`\`feat`\`fix` in the any of commit's message (here some explanations how logic will determine what to increment if be 3 commits with all 3 options).
2. Image for DEV: ....
3. Image for PROD: After merge to Master...
etc.
```

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        TAG=$([[ "$GITHUB_REF" == "refs/heads/master" ]] && echo "${{ steps.versioning.outputs.tag }},latest" || echo "${{ steps.versioning.outputs.tag }}")
+        EOF
+
+    - name: List environment info
+      run: |
+        make show-env
+        make info
+
+    - name: Set up QEMU
+      uses: docker/setup-qemu-action@v1
```
Comment (devProdigy) [inline .github/workflows/docker.yml:82]: could you please explain why is this needed?

File: .github/workflows/docker.yml [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        REGISTRY: ${{ secrets.DOCKER_REGISTRY }}
+        REPO: ${{ secrets.DOCKER_REPO }}
+      with:
+        context: .
+        file: ./Dockerfile
+        tags: |
+          ${{ env.BASE_IMAGE }}
+          ${{ env.BASE_IMAGE }}:${{ env.VERSION }}
+        labels: |
+          ${{ steps.docker_meta.outputs.labels }}
```
Comment (o-fedorov) [inline .github/workflows/docker.yml:105]: Is it really needed?

**Review by o-fedorov — DISMISSED** [review]

File: .github/workflows/docker.yml (hunk 1 above)
Comment (RollerKnobster) [reply .github/workflows/docker.yml:6]: I'll comment out the `pull_request` part then and leave a TODO to bring it back when Gordo client is separated.

File: README.md (hunk 2 above)
Comment (RollerKnobster) [reply README.md:241]: I meant the ones in the beginning of the section. Will clarify.

File: README.md (hunk 3 above)
Comment (RollerKnobster) [reply README.md:260]: Will elaborate.

File: .github/workflows/docker.yml (hunk 4 above)
Comment (RollerKnobster) [reply .github/workflows/docker.yml:82]: To be honest, we probably don't need QEMU. `buildx` is needed in order to enable the `build-push-action` to function properly, as with moving to `v2` it moved to using `docker buildx` toolset for building images. But we certainly don't need support for multiple platforms. Will remove.

File: .github/workflows/docker.yml (hunk 5 above)
Comment (RollerKnobster) [reply .github/workflows/docker.yml:105]: I guess we could do it all manually. This is needed to automate at least some part of the standard set of [OCI labels](https://github.com/opencontainers/image-spec/blob/master/annotations.md), but we could do without that.

File: .github/workflows/docker.yml (hunk 1 above)
Comment (devProdigy) [reply .github/workflows/docker.yml:7]: did you comment it?)

File: .github/workflows/docker.yml [hunk 6]
```diff
@@ -0,0 +1,170 @@
+name: Docker
+
+on:
+#  We should try and build the image on every pull request, but only after we clear out the Gordo dependencies
```
Comment (o-fedorov) [inline .github/workflows/docker.yml:4]: Could you please create an Azure DevOps task for it (if there is no already) and place a link or ID of it here?

File: .github/workflows/docker.yml [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        docker pull -q alpine > /dev/null
+        docker run --rm -v "$(pwd)/trivydb/":/root/.cache/ $(docker images alpine --format '{{.Repository}}:{{.Tag}}' | head -1) chmod 777 -R /root/.cache/
+
+    - name: Push images to dev registry
+      if: ${{ !startsWith(github.ref , 'refs/pull/') }}
+      env:
+        USERNAME: ${{ secrets.DOCKER_USERNAME }}
+        PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
+        REGISTRY: ${{ secrets.DOCKER_REGISTRY }}
+        REPO: ${{ secrets.DOCKER_REPO }}
```
Comment (o-fedorov) [inline .github/workflows/docker.yml:140]: For better readability it's better to have REGISTRY and REPO configured in this file, not in secrets.

BTW, now that we've had problems with docker registry credentials multiple times I start to thinking of exposing DOCKER_USERNAME as well, so that when credentials should be updated it's easier to find the client that has access to the registry. That is, I think only `DOCKER_PASSWORD` and `PROD_DOCKER_PASSWORD` should be in the secrets.

Still, it's not a requirement for current PR.

**Review by o-fedorov — DISMISSED** [review]

File: .github/workflows/docker.yml (hunk 7 above)
Comment (RollerKnobster) [reply .github/workflows/docker.yml:140]: I am all for that actually.

File: .github/workflows/docker.yml (hunk 6 above)
Comment (RollerKnobster) [reply .github/workflows/docker.yml:4]: Absolutely.

**Review by devProdigy — APPROVED** [review]
